### ISSUE TITLE
Aggregate Stage 1 oracle metrics instead of leaving W&B summary at the last batch

### ISSUE THREAD

**Issue by tokenbender** [mananchawla2005/adaptive-tokenization#91]
## Summary

The public Phase 2 Stage 1 oracle run exposes `oracle/best_cr`, `oracle/best_loss`, and `oracle/best_reward` as run-summary metrics, but the Stage 1 code logs those names once per mini-batch and never computes a dataset-level aggregate. Since the final explicit summary only writes `oracle_num_examples` and `oracle_k_samples`, the public scalar is not established as a statistic over the 1000 generated oracle examples.

This is a reporting/scientific-validation bug: a reader can reasonably read `oracle/best_loss=2.4413` beside `oracle_num_examples=1000` as the oracle dataset's aggregate best loss, but the implementation only proves it was a logged mini-batch mean under that metric name.

This is distinct from #27 and #84. #27 covers the oracle target selection objective. #84 covers the unsupported README claim of Phase 2 oracle CORE=0.951. This issue is only about the semantics of the public Stage 1 oracle scalar summaries.

## Evidence in this repo

Current commit checked: `d53f400df338eee48641df72ea9aad2a6950fd14`.

1. Phase 2 Stage 1 creates the public run with `num_prompts=1000` and `k_samples=128`:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/phase2_app.py#L86-L96

2. `stage1_create_oracle` iterates a dataloader with `batch_size=2`:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/src/hybrid_train.py#L140

3. After each batch, it computes mean best compression/loss/reward for that batch and logs them with the dataset-level-looking names `oracle/best_cr`, `oracle/best_loss`, and `oracle/best_reward`:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/src/hybrid_train.py#L187-L195

   ```python
   best_cr_mean = cr[best_idx, torch.arange(B)].mean().item()
   best_loss = losses_t[best_idx, torch.arange(B)].mean().item()
   best_reward = rewards[best_idx, torch.arange(B)].mean().item()
   tracker.log({
       "oracle/best_cr": best_cr_mean,
       "oracle/best_loss": best_loss,
       "oracle/best_reward": best_reward,
   }, step=collected)
   ```

4. At the end, the only explicit W&B summary keys are counts/config values. There is no accumulated sum/count, mean, median, standard deviation, histogram, or saved per-example table for `oracle/best_*`:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/src/hybrid_train.py#L223-L227

5. The local tracker wrapper confirms the split: `log()` forwards to `wandb.log`, while `summary()` only writes the keys explicitly provided:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/src/tracker.py#L75-L82

## Public W&B evidence

The public run is:

- https://wandb.ai/highattacker/adaptive-tokenization/runs/xksxw9h6

Unauthenticated W&B GraphQL summary for that run currently exposes:

```text
displayName          exp-20260623-131726-gd53f400-phase2-stage1-oracle
_step                1000
oracle/best_cr       0.4377126097679138
oracle/best_loss     2.4412994384765625
oracle/best_reward  -2.247549295425415
oracle_k_samples     128
oracle_num_examples  1000
```

That public surface makes the `oracle/best_*` values look like run-level oracle results. The code path above does not establish that interpretation; it only establishes per-mini-batch logged values, with the final summary explicitly setting only count and sample-count keys.

## Why this is a bug / expected behavior

A reported oracle dataset metric should either be:

- reduced over the examples it claims to summarize, or
- named as a batch/local diagnostic, for example `oracle/batch_best_loss`.

Here the metric name and public summary imply a dataset statistic, but the code never performs the corresponding reduction. That is enough to make the current public value scientifically ambiguous and potentially misleading.

The expected behavior is to accumulate per-example or token-weighted numerators/denominators across all 1000 collected prompts, then write explicit summary keys such as:

```text
oracle/best_loss_mean
oracle/best_loss_median
oracle/best_loss_std
oracle/best_cr_mean
oracle/best_reward_mean
oracle/num_examples
```

or to keep the current logging but rename it to `oracle/batch_best_*` and avoid using it as evidence for oracle quality.

## Suspected culprit

The culprit is the per-batch `tracker.log()` block in `stage1_create_oracle`:

- https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/src/hybrid_train.py#L187-L195

It uses final-result-looking metric names for batch means and does not maintain `sum`, `count`, or per-example records. The end-of-run `tracker.summary()` block then omits aggregate oracle quality metrics, leaving the logged scalar summary to stand in for something it did not compute.

### nanochat

nanochat's BPB evaluation accumulates `total_nats` and `total_bytes` across evaluation steps, reduces them across ranks, then divides once to produce the reported metric:

- https://github.com/karpathy/nanochat/blob/dc54a1a3077cab11d68fac4c5d1cd5c51f5d8c7a/nanochat/loss_eval.py#L27-L65

nanochat also writes evaluation results to durable CSV/report outputs rather than relying on an ambiguous final logged batch value:

- https://github.com/karpathy/nanochat/blob/dc54a1a3077cab11d68fac4c5d1cd5c51f5d8c7a/scripts/base_eval.py#L285-L317

### TorchTitan

TorchTitan's validator counts valid tokens, accumulates validation losses across validation batches, averages before logging, and logs the reduced validation loss:

- https://github.com/pytorch/torchtitan/blob/6392a5da3a349cb402ed8bc2a063382abe2e6618/torchtitan/components/validate.py#L249-L317

TorchTitan's RL metrics layer also makes the reduction contract explicit: `Mean.reduce` sums values and counts, and the metrics processor emits reduced keys after aggregation:

- https://github.com/pytorch/torchtitan/blob/6392a5da3a349cb402ed8bc2a063382abe2e6618/torchtitan/experiments/rl/observability/metrics/types.py#L137-L152
- https://github.com/pytorch/torchtitan/blob/6392a5da3a349cb402ed8bc2a063382abe2e6618/torchtitan/experiments/rl/observability/metrics/processor.py#L114-L156

These references establish the expected behavior: run-level evaluation/oracle metrics should be explicit reductions over the evaluated population, or the metric name should say that it is a local/batch diagnostic.

## Suggested fix

- During Stage 1, store per-example selected `loss`, `cr`, and `reward` alongside the selected boundaries, or accumulate sums/counts online.
- At the end of Stage 1, compute and write explicit aggregate summary metrics.
- Optionally log a W&B table or JSON/CSV artifact with per-example `prompt_id`, selected candidate index, loss, compression ratio, reward, seed, and prompt/answer lengths.
- Rename the existing live progress keys to `oracle/batch_best_*` if they remain per-batch diagnostics.

## Expected impact

The current public Stage 1 oracle summary may overstate the precision of the oracle evidence. It weakens Phase 2 claims that rely on oracle quality or headroom because the visible `oracle/best_*` scalars are not proven to summarize the full generated oracle dataset.
